README.md
# deno-steam-totp

This is a deno version of [node-steam-totp by DoctorMcKay](https://github.com/DoctorMcKay/node-steam-totp)

It's a port of his module, just a little bit different.

## usage

```js
import * as SteamTotp from 'https://deno.land/x/steamtotp@v3.0.1';
```

## generateAuthCode( secret: string | Buffer, timeOffset?: number): string

Generate a Steam-style TOTP authentication code.

## generateConfirmationCode (identitySecret: Buffer | string, options: Options): string

Generate a base64 confirmation key for use with mobile trade confirmations. The code can only be used once.

### options

```ts
options?: {
    time?: number;
    tag?: "conf" | "details" | "allow" | "cancel";
}
```

## getDeviceID(steamID: string | object, salt?: string): string

Get a standardized device ID based on your SteamID.

### steamID - Your SteamID, either as a string or as an object which has a toString() method that returns the SteamID

`STEAM_TOTP_SALT` is not supported.

The algorithm used is:

1. Convert the SteamID to a string
2. Append the value of the `salt` param to the SteamID, If It's set
3. SHA-1 hash it and encode the resulting hash as a lowercase hex value
4. Truncate the hash to 32 characters
5. Insert dashes such that the resulting value has 5 groups of hexadecimal values containing 8, 4, 4, 4, and 12 characters, respectively
6. Prepend "android:" to the resulting value

## getLocalUnixTime(offset = 0): number

Returns the current local Unix time

## getTimeOffset(): Promise&lt;number&gt;

Returns offset between you and steam servers in seconds


CHANGELOG.md
# Changelog

## v3.0.1

- allow any string as tag in `generateConfirmationKey` options

## v3.0.0

- rename `generateConfirmationCode` to `generateConfirmationKey`

## v2.1.0

- export time functions


mod.ts
import { Buffer, HmacSha1, Sha1 } from "./deps.ts";
import { getLocalUnixTime } from "./time.ts";

export * from "./time.ts";

/**
 * Generate a Steam-style TOTP authentication code.
 * @param sharedSecret - Your TOTP shared_secret as a Buffer, hex, or base64
 * @param timeOffset - If you know how far off your clock is from the Steam servers, put the offset here in seconds.
 */
export function generateAuthCode(
  sharedSecret: string | Buffer,
  timeOffset?: number,
): string {
  const secretBuffer = bufferizeSecret(sharedSecret);

  const offsetedLocalUnixTime = getLocalUnixTime(timeOffset || 0);

  const timeBuffer = Buffer.allocUnsafe(8);
  timeBuffer.writeUInt32BE(0, 0);
  timeBuffer.writeUInt32BE(Math.floor(offsetedLocalUnixTime / 30), 4);

  let hmac = Buffer.from(
    new HmacSha1(secretBuffer).update(timeBuffer).digest(),
  );

  const start = hmac[19] & 0x0F;
  hmac = hmac.slice(start, start + 4);

  let fullcode = hmac.readUInt32BE(0) & 0x7FFFFFFF;

  const chars = "23456789BCDFGHJKMNPQRTVWXY";

  let code = "";
  for (let i = 0; i < 5; i++) {
    code += chars.charAt(fullcode % chars.length);
    fullcode /= chars.length;
  }

  return code;
}

/**
 * Generate a base64 confirmation key for use with mobile trade confirmations. The code can only be used once.
 * @param identitySecret - The identity_secret that you received when enabling two-factor authentication
 * @param  time - The Unix time for which you are generating this secret. Generally should be the current time.
 * @param tag - Default: "conf".
 * The tag which identifies what this request (and therefore key) will be for.
 * "conf" to load the confirmations page.
 * "details" to load details about a trade.
 * "allow" to confirm a trade.
 * "cancel" to cancel it.
 */
export function generateConfirmationKey(
  identitySecret: Buffer | string,
  options?: {
    time?: number;
    /** "conf" | "details" | "allow" | "cancel" */
    tag?: string;
  },
): string {
  if (!options) {
    options = {};
  }

  const tag = options.tag || "conf";

  const identitySecretBuffer = bufferizeSecret(identitySecret);

  let dataLen = 8;

  if (tag) {
    if (tag.length > 32) {
      dataLen += 32;
    } else {
      dataLen += tag.length;
    }
  }

  // deno-lint-ignore prefer-const
  let paramBuffer = Buffer.allocUnsafe(dataLen);
  paramBuffer.writeUInt32BE(0, 0);
  paramBuffer.writeUInt32BE(
    options.time !== undefined ? options.time : getLocalUnixTime(),
    4,
  );

  if (tag) {
    paramBuffer.write(tag, 8);
  }
  return Buffer.from(
    new HmacSha1(identitySecretBuffer).update(paramBuffer).digest(),
  ).toString("base64");
}

export function bufferizeSecret(secret: string | Buffer): Buffer {
  if (typeof secret === "string") {
    // Check if it's hex
    if (secret.match(/[0-9a-f]{40}/i)) {
      return Buffer.from(secret, "hex");
    } else {
      // Looks like it's base64
      return Buffer.from(secret, "base64");
    }
  }
  return secret;
}

/**
 * Get a standardized device ID based on your SteamID.
 * @param steamID - Your SteamID, either as a string or as an object which has a toString() method that returns the SteamID
 */
// deno-lint-ignore ban-types
export function getDeviceID(steamID: string | object, salt?: string): string {
  const steamIDString = typeof steamID === "string"
    ? steamID
    : steamID.toString();
  return "android:" +
    new Sha1().update(steamIDString + (salt || "")).hex().replace(
      /^([0-9a-f]{8})([0-9a-f]{4})([0-9a-f]{4})([0-9a-f]{4})([0-9a-f]{12}).*$/,
      "$1-$2-$3-$4-$5",
    );
}


time.ts
/**
 * Returns the current local Unix time
 * @param offset - This many seconds will be added to the returned time
 */
export function getLocalUnixTime(offset?: number): number {
  return Math.floor(Date.now() / 1000) + (offset || 0);
}

/**
 * Returns offset between you and steam servers in seconds
 * 
 * Offset will be fetched from: https://api.steampowered.com/ITwoFactorService/QueryTime/v1/
 */
export async function getTimeOffset(): Promise<number> {
  const response = await fetch(
    "https://api.steampowered.com/ITwoFactorService/QueryTime/v1/",
    {
      method: "POST",
      headers: {
        "Content-Length": "0",
      },
    },
  )
    .then((res) => res.json())
    .then((res) => res.response);

  if (!response || !response.server_time) {
    throw new Error("Malformed response");
  }

  const offset = response.server_time - getLocalUnixTime();

  return offset;
}


main.ts
import { Buffer, HmacSha1, Sha1 } from "./deps.ts";
import { getLocalUnixTime, getTimeOffset } from "./time.ts";

/**
 * Generate a Steam-style TOTP authentication code.
 * @param sharedSecret - Your TOTP shared_secret as a Buffer, hex, or base64
 * @param timeOffset - If you know how far off your clock is from the Steam servers, put the offset here in seconds.
 */
function generateAuthCode(
  sharedSecret: string | Buffer,
  timeOffset?: number,
): string {
  const secretBuffer = bufferizeSecret(sharedSecret);

  const offsetedLocalUnixTime = getLocalUnixTime(timeOffset || 0);

  const timeBuffer = Buffer.allocUnsafe(8);
  timeBuffer.writeUInt32BE(0, 0);
  timeBuffer.writeUInt32BE(Math.floor(offsetedLocalUnixTime / 30), 4);

  let hmac = Buffer.from(
    new HmacSha1(secretBuffer).update(timeBuffer).digest(),
  );

  const start = hmac[19] & 0x0F;
  hmac = hmac.slice(start, start + 4);

  let fullcode = hmac.readUInt32BE(0) & 0x7FFFFFFF;

  const chars = "23456789BCDFGHJKMNPQRTVWXY";

  let code = "";
  for (let i = 0; i < 5; i++) {
    code += chars.charAt(fullcode % chars.length);
    fullcode /= chars.length;
  }

  return code;
}

/**
 * Generate a base64 confirmation key for use with mobile trade confirmations. The code can only be used once.
 * @param identitySecret - The identity_secret that you received when enabling two-factor authentication
 * @param  time - The Unix time for which you are generating this secret. Generally should be the current time.
 * @param tag - Default: "conf".
 * The tag which identifies what this request (and therefore key) will be for.
 * "conf" to load the confirmations page.
 * "details" to load details about a trade.
 * "allow" to confirm a trade.
 * "cancel" to cancel it.
 */
function generateConfirmationCode(
  identitySecret: Buffer | string,
  options?: {
    time?: number;
    tag?: "conf" | "details" | "allow" | "cancel";
  },
): string {
  if (!options) {
    options = {};
  }

  const tag = options.tag || "conf";

  const identitySecretBuffer = bufferizeSecret(identitySecret);

  let dataLen = 8;

  if (tag) {
    if (tag.length > 32) {
      dataLen += 32;
    } else {
      dataLen += tag.length;
    }
  }

  // deno-lint-ignore prefer-const
  let paramBuffer = Buffer.allocUnsafe(dataLen);
  paramBuffer.writeUInt32BE(0, 0);
  paramBuffer.writeUInt32BE(
    options.time !== undefined ? options.time : getLocalUnixTime(),
    4,
  );

  if (tag) {
    paramBuffer.write(tag, 8);
  }
  return Buffer.from(
    new HmacSha1(identitySecretBuffer).update(paramBuffer).digest(),
  ).toString("base64");
}

function bufferizeSecret(secret: string | Buffer): Buffer {
  if (typeof secret === "string") {
    // Check if it's hex
    if (secret.match(/[0-9a-f]{40}/i)) {
      return Buffer.from(secret, "hex");
    } else {
      // Looks like it's base64
      return Buffer.from(secret, "base64");
    }
  }
  return secret;
}

/**
 * Get a standardized device ID based on your SteamID.
 * @param steamID - Your SteamID, either as a string or as an object which has a toString() method that returns the SteamID
 */
// deno-lint-ignore ban-types
function getDeviceID(steamID: string | object): string {
  const salt = Deno.env.get("STEAM_TOTP_SALT") || "";
  const steamIDString = typeof steamID === "string"
    ? steamID
    : steamID.toString();
  return "android:" + new Sha1().update(steamIDString + salt).hex().replace(
    /^([0-9a-f]{8})([0-9a-f]{4})([0-9a-f]{4})([0-9a-f]{4})([0-9a-f]{12}).*$/,
    "$1-$2-$3-$4-$5",
  );
}

export {
  generateAuthCode,
  generateConfirmationCode,
  getDeviceID,
  getLocalUnixTime,
  getTimeOffset,
};
